[inductor] Switch inductor_prims._bucketize over to aten.bucketize

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #106658

inductor_prims._bucketize was added while we worked on hardening the inductor lowering. Now the lowering should be sufficiently tested and should have good enough perf (https://github.com/pytorch/pytorch/pull/104456) - so we can remove the temporary `inductor_prims._bucketize` op and move the lowerings to the `aten.bucketize` op.

Note that we haven't added a CPU implementation yet.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

davidberard98 committed:
[WIP][inductor] Switch inductor_prims._bucketize over to aten.bucketize

inductor_prims._bucketize was added while we worked on hardening the inductor lowering. Now the lowering should be sufficiently tested and should have sufficient perf so we can remove the temporary `inductor_prims._bucketize` op and move the lowerings to the `aten.bucketize` op

[ghstack-poisoned]

eellison reviewed (approved):
nice!

davidberard98 commented:
@pytorchbot merge

davidberard98 commented:
@pytorchbot merge

aakhundov commented on test/inductor/test_torchinductor_codegen_dynamic_shapes.py:
```diff
@@ -160,9 +160,9 @@ def run(*ex, **kwargs):
-    "test_inductor_bucketize_dynamic_shapes": TestFailure("cpu"),
-    "test_inductor_bucketize_default_kwargs_dynamic_shapes": TestFailure("cpu"),
-    "test_inductor_bucketize_int_dynamic_shapes": TestFailure("cpu"),
+    "test_bucketize_dynamic_shapes": TestFailure("cpu"),
+    "test_bucketize_default_kwargs_dynamic_shapes": TestFailure("cpu"),
+    "test_bucketize_int_dynamic_shapes": TestFailure("cpu"),
```
I wonder, why we don't need to mention `test_bucketize_add_autotune` and `test_bucketize_computed_offsets` here. Are those tests not attempted to run on CPU?

davidberard98 replied:
The add/subtract gets compiled in that case so it ends up passing these tests. IMO it's fine to leave them un-skipped, even if they're passing for kind of the wrong reason

aakhundov commented on torch/_inductor/lowering.py:
```diff
@@ -1627,8 +1627,8 @@ def inner_fn(index):
 
-@register_lowering(inductor_prims._bucketize, type_promotion_kind=None)
-def _inductor_bucketize(
+@register_lowering(aten.bucketize, type_promotion_kind=None)
```
Out of curiosity: do we have data on the current Triton lowering's perf compared to `aten.bucketize`?

davidberard98 replied:
In the comments of #104456 we're showing better performance than aten.bucketize (.30ms eager, .22ms PT2), on a specific example (2^24 float32 values, 1024 buckets)

davidberard98 committed:
Update on "[inductor] Switch inductor_prims._bucketize over to aten.bucketize"


inductor_prims._bucketize was added while we worked on hardening the inductor lowering. Now the lowering should be sufficiently tested and should have good enough perf (https://github.com/pytorch/pytorch/pull/104456) - so we can remove the temporary `inductor_prims._bucketize` op and move the lowerings to the `aten.bucketize` op.

Note that we haven't added a CPU implementation yet.

cc voznesenskym penguinwu @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot merge
